### PR TITLE
[WIP] Feat/decouple sympc tests from pysyft

### LINKED ISSUE #7: Simulate the operations on the Workers
To properly test SyMPC we need to have installed the syft_0.3.0.

Add a method to simulate the parties operations (send and get) such that we might have a simple way to test some of the operations from our repository without installing syft.

### PR DESCRIPTION
**Pull request by danielorihuela**
## Description
Decouple SyMPC tests from PySyft `Module` class.

Related to #7.

## How has this been tested?
- Already existent tests, which now use a different "Module".

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A). No because I cannot
- [x] My changes are covered by tests

### CODE REVIEW

**Comment by danielorihuela** [comment]
I decided to put the Module Stub in a tests/stubs/module_stub. Tests fails because all files must follow the pattern *_test. Where should I put the stub?

**Comment by gmuraru** [comment]
But we still have the "workers" coupling problem, right? (because we use `sy.VirtualMachine`) and the `send` + `get` methods.

With this you decouple only the Module, right?

**Comment by gmuraru** [comment]
Ahh! I talked with @tudorcebere and for having a better `decoupling` (not for the workers) we could also move the `allowlist` in SyMPC, from Syft. creating now an issue for that

**Comment by gmuraru** [comment]
Did you run those changes with the Duet Notebooks from Syft (we have the notebooks in the ```examples``` folder)?

**Comment by danielorihuela** [comment]
1. True, I am going to put this as WIP and keep working.
2. No, I did not run those tests. I only created a Module stub so they should keep working. Anyway, I still have work to do with the VirtualMachines as you said.

Thanks!

**Comment by danielorihuela** [comment]
I double checked on the Duet network MNIST example after doing `pip install .` on the sympc folder. It all keeps working.

Your question: "Did you run those changes with the Duet Notebooks from Syft (we have the notebooks in the examples folder)?". Makes me feel like we are not on the same page. I think that I did not explain myself well enough.
What I am doing at the moment is creating fake classes for the tests. So, SyMPC will still be dependent on the PySyft on its source code. Nevertheless, SyMPC tests will no longer be dependent on PySyft code.

SyMPC (src) will still be dependent on PySyft.
SyMPC (tests) will *NOT* depend on PySyft code anymore.

I think that's the easier way to remove the Circular dependency kind of error. I am aware that this does not decouples SyMPC from PySyft completely. So SyMPC will not be a standalone yet, which I understand that is the end goal. However, doing all that in one step seemed like to much work to be done at once. For that reason I am focusing on decoupling the tests. To make PySyft and SyMPC development easier.

File: stubs/module_stub.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
+"""File containing a stub of the Module that can be found in the PySyft library.
```
Comment (gmuraru) [inline stubs/module_stub.py:1]: Really like the explanations!

**Comment by gmuraru** [comment]
> I double checked on the Duet network MNIST example after doing `pip install .` on the sympc folder. It all keeps working.
> 
> Your question: "Did you run those changes with the Duet Notebooks from Syft (we have the notebooks in the examples folder)?". Makes me feel like we are not on the same page. I think that I did not explain myself well enough.
> What I am doing at the moment is creating fake classes for the tests. So, SyMPC will still be dependent on the PySyft on its source code. Nevertheless, SyMPC tests will no longer be dependent on PySyft code.
> 
> SyMPC (src) will still be dependent on PySyft.
> SyMPC (tests) will _NOT_ depend on PySyft code anymore.
> 
> I think that's the easier way to remove the Circular dependency kind of error. I am aware that this does not decouples SyMPC from PySyft completely. So SyMPC will not be a standalone yet, which I understand that is the end goal. However, doing all that in one step seemed like to much work to be done at once. For that reason I am focusing on decoupling the tests. To make PySyft and SyMPC development easier.
> 
> If you are okey with that reasoning. Then the only thing left to do would be to "copy" the VirtualMachine into SyMPC and remove all the unwanted code. Building a fake VirtualMachine with the minim amount of code that makes tests work.

Yep! you are right that we should do this step by step! Maybe we can merge this PR after getting the other logic decoupling PRs first.

The question regarding the notebook, in Syft, we have some Duet notebooks specific to SyMPC - my question was more targeted if you tested those notebooks - not the Syft specific ones (we have there a notebook where we do Inference using Duet)

File: stubs/module_stub.py (hunk 1 above)
Comment (danielorihuela) [reply stubs/module_stub.py:1]: Thanks

**Comment by danielorihuela** [comment]
> > I double checked on the Duet network MNIST example after doing `pip install .` on the sympc folder. It all keeps working.
> > Your question: "Did you run those changes with the Duet Notebooks from Syft (we have the notebooks in the examples folder)?". Makes me feel like we are not on the same page. I think that I did not explain myself well enough.
> > What I am doing at the moment is creating fake classes for the tests. So, SyMPC will still be dependent on the PySyft on its source code. Nevertheless, SyMPC tests will no longer be dependent on PySyft code.
> > SyMPC (src) will still be dependent on PySyft.
> > SyMPC (tests) will _NOT_ depend on PySyft code anymore.
> > I think that's the easier way to remove the Circular dependency kind of error. I am aware that this does not decouples SyMPC from PySyft completely. So SyMPC will not be a standalone yet, which I understand that is the end goal. However, doing all that in one step seemed like to much work to be done at once. For that reason I am focusing on decoupling the tests. To make PySyft and SyMPC development easier.
> > If you are okey with that reasoning. Then the only thing left to do would be to "copy" the VirtualMachine into SyMPC and remove all the unwanted code. Building a fake VirtualMachine with the minim amount of code that makes tests work.
> 
> Yep! you are right that we should do this step by step! Maybe we can merge this PR after getting the other logic decoupling PRs first.
> 
> The question regarding the notebook, in Syft, we have some Duet notebooks specific to SyMPC - my question was more targeted if you tested those notebooks - not the Syft specific ones (we have there a notebook where we do Inference using Duet)

1. Okey, I will continue working on the "decoupling" of the VirtualMachine, that I saw has a lot of code involved due to inheritance and other classes involder. VirtualMachine -> Node -> AbstractNode -> Address -> Serializable. VirtualServerConnection is also involder wihch inherits from ServerConnection. If no new member wants to solve #155 I will take it.
2. PySyft has a lot of notebooks. I cannot seem to find the PySyft notebooks targeting SyMPC.

**Comment by danielorihuela** [comment]
Stubing the VirtualMachine is difficult with my current knowledge of the system. There are to much code involved. Only the VirtualMachine has 5 levels of inheritance (VirtualMachine -> Node -> AbstractNode -> Address -> Serializable). VirtualMachine calls functions that again, involved multiple levels of inheritance.

I partially stubbed the VirtualMachine but I get the following error.
image
image

Furthermore, when trying to implement the VirtualMachine, the tests that worked with the Module stub started failing. I do not have the required knowledge of the whole system.
For that reason I am closing this.

**Comment by gmuraru** [comment]
I think we can "park" this issue for the moment - we need to think at a higher-level general API that should call the lower-level specific API (in our case Syft)

**Pull request closed without merging**